### ISSUE TITLE
Method References on object instances or field accesses parsed to TypeExpr instead of expected NameExpr or FieldAccessExpr

### ISSUE THREAD

**Issue by scootafew** [javaparser/javaparser#2455]
Since the addition of a `resolve` method to `MethodReferenceExpr` I've been trying to resolve some different types of method references. If we take the example of a call to `System.out::println`, when it is expressed as a method reference it cannot be resolved (instead throws `UnsupportedOperationException`), whilst the equivalent regular method call is resolved with no problems.

So this test passes:
```
@Test
    void resolveMethodCallExpr() {
        JavaParser parser = createParserWithResolver(new ReflectionTypeSolver());
        MethodCallExpr methodCall = parser.parse("class X{void x(){System.out.println(\"Hello world!\");}}")
                .getResult().get()
                .findFirst(MethodCallExpr.class).get();
        ResolvedMethodDeclaration resolvedMethodRef = methodCall.resolve();
        assertEquals("println", resolvedMethodRef.getName());
        assertEquals("void", resolvedMethodRef.getReturnType().describe());
    }
```

But this one does not:
```
    @Test
    void resolveMethodReferenceExpr() {
        JavaParser parser = createParserWithResolver(new ReflectionTypeSolver());
        MethodReferenceExpr methodRef = parser.parse("class X{void x(){Consumer<? super String>r=System.out::println;}}")
                .getResult().get()
                .findFirst(MethodReferenceExpr.class).get();
        ResolvedMethodDeclaration resolvedMethodRef = methodRef.resolve();
        assertEquals("println", resolvedMethodRef.getName());
        assertEquals("void", resolvedMethodRef.getReturnType().describe());
    }
```

From the debugging I've done and the comments made in #2210 this does seem to be down to the way method references are being parsed. In the examples above, the scope of the call is `System.out`, which for a method reference is parsed to a `ClassOrInterfaceType` wrapped in a `TypeExpr`. The equivalent for a method call is parsed to a `FieldAccessExpr` which has the correct symbol-solver logic.

I'm hoping someone could please shed some light on where this behavior would be defined (I'm hoping not in JavaCC :P)

**Comment by matozoid** [comment]
The AST seems reasonable to me (correct me if I'm wrong - I often am :-) )
System is a type
out is a name

In JavaParserFacade line 385, the solver is rather dumb. It grabs `id`, which is `out`, and looks for the its declaration, completely ignoring `System`. What it should have done is solve the whole name expression!

**Comment by jlerbsc** [comment]
I made the same observation and i found that TypeExtractor was doing something else when it tried to resolve fieldAccessExpr.

    public ResolvedType visit(FieldAccessExpr node, Boolean solveLambdas) {
        // We should understand if this is a static access
        if (node.getScope() instanceof NameExpr ||
                node.getScope() instanceof FieldAccessExpr) {
            Expression staticValue = node.getScope();
            SymbolReference<ResolvedTypeDeclaration> typeAccessedStatically = JavaParserFactory.getContext(node, typeSolver).solveType(staticValue.toString());
     ...

**Comment by scootafew** [comment]
Yeah so I had also seen the same and decided to try just changing where `JavaParserFacade` uses `out` to instead use `System.out` or even just `System` in case that resulted in different behaviour but it did not. As @jlerbsc rightly pointed out the symbol solver needs to use logic as in `TypeExtractor`, which incidentally is what it does when solving a `MethodCallExpr` to `System.out` (which in this case is a fieldAccessExpr).

![image](https://user-images.githubusercontent.com/20837812/70952487-47b0b880-205e-11ea-98b2-59dc71249af3.png)


![image](https://user-images.githubusercontent.com/20837812/70952261-97db4b00-205d-11ea-9379-4cf8e7451609.png)

Please likewise feel free to correct me, I'm just trying to figure out why the same thing is parsed differently in these two cases!

**Comment by scootafew** [comment]
Found another example test case that fails:

```
@Test
    void resolveMethodReferenceExpr3() {
        JavaParser parser = createParserWithResolver(new ReflectionTypeSolver());
        CompilationUnit cu = parser.parse("import java.util.ArrayList; class X{ArrayList<Integer> a = new ArrayList<>(); void x(){ Consumer<? super Integer> r = a::add;}}")
                .getResult().get();
        System.out.println(cu);
        MethodReferenceExpr methodRef = cu.findFirst(MethodReferenceExpr.class).get();
        ResolvedMethodDeclaration resolvedMethodRef = methodRef.resolve();
        assertEquals("add", resolvedMethodRef.getName());
        assertEquals("void", resolvedMethodRef.getReturnType().describe());
    }
```
![image](https://user-images.githubusercontent.com/20837812/71013466-a4f14c00-20e8-11ea-8a7e-52d2bfb84441.png)

You can see above how `a` is once again a `ClassOrInterfaceType` wrapped in a `TypeExpr` in the AST, whereas I would have expected it to be a `NameExpr`?

**Comment by matozoid** [comment]
I bet that's coming from here: https://github.com/javaparser/javaparser/blob/master/javaparser-core/src/main/javacc/java.jj#L1849-L1855

Not saying that it is wrong, I'm too tired to figure it out right now :-#

**Comment by scootafew** [comment]
I know the feeling! Thanks for the pointer, I've been having a look at the grammar & the language spec but I'm not really confident that it's right or wrong either. I have noticed that when a method reference is of the format `this::doSomething` the scope is a `ThisExpr` so I might try and fix the symbol solving for that case as a starting point

**Comment by jlerbsc** [comment]
This issue appears to have been fixed in the current version. However, there is an import missing in the example provided that did not seem to work originally. 

    @Test
    void resolveMethodReferenceExpr() {
        String code = "import java.util.function.Consumer;\n"
                + "class X{void x(){Consumer<? super String>r=System.out::println;}}";

        JavaParser parser = createParserWithResolver(new ReflectionTypeSolver());
        MethodReferenceExpr methodRef = parser.parse(code)
                .getResult().get()
                .findFirst(MethodReferenceExpr.class).get();
        ResolvedMethodDeclaration resolvedMethodRef = methodRef.resolve();
        assertEquals("println", resolvedMethodRef.getName());
        assertEquals("void", resolvedMethodRef.getReturnType().describe());
    }
